src/main/java/io/airlift/compress/zstd/ZstdIncrementalFrameDecompressor.java

```diff
@@ -0,0 +1,390 @@
+import static io.airlift.compress.zstd.UnsafeUtil.UNSAFE;
+import static io.airlift.compress.zstd.Util.checkArgument;
+import static io.airlift.compress.zstd.Util.checkState;
+import static io.airlift.compress.zstd.Util.fail;
+import static io.airlift.compress.zstd.Util.verify;
+import static io.airlift.compress.zstd.ZstdFrameDecompressor.MAX_WINDOW_SIZE;
+import static io.airlift.compress.zstd.ZstdFrameDecompressor.decodeRawBlock;
+import static io.airlift.compress.zstd.ZstdFrameDecompressor.decodeRleBlock;
+import static io.airlift.compress.zstd.ZstdFrameDecompressor.readFrameHeader;
+import static io.airlift.compress.zstd.ZstdFrameDecompressor.verifyMagic;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
+public class ZstdIncrementalFrameDecompressor
+{
+    private enum State {
+        INITIAL,
+        READ_FRAME_MAGIC,
+        READ_FRAME_HEADER,
+        READ_BLOCK_HEADER,
+        READ_BLOCK,
+        READ_BLOCK_CHECKSUM,
+        FLUSH_OUTPUT
+    }
+
+    private final ZstdFrameDecompressor frameDecompressor = new ZstdFrameDecompressor();
+
+    private State state = State.INITIAL;
+    private FrameHeader frameHeader;
+    private int blockHeader = -1;
+
+    private int inputConsumed;
+    private int outputBufferUsed;
+
+    private int inputRequired;
+    private int requestedOutputSize;
+
+    // current window buffer
+    private byte[] windowBase = new byte[0];
+    private long windowAddress = ARRAY_BYTE_BASE_OFFSET;
+    private long windowLimit = ARRAY_BYTE_BASE_OFFSET;
+    private long windowPosition = ARRAY_BYTE_BASE_OFFSET;
+
+    private XxHash64 partialHash;
+
+    public boolean isAtStoppingPoint()
+    {
+        return state == State.READ_FRAME_MAGIC;
+    }
+
+    public int getInputConsumed()
+    {
+        return inputConsumed;
+    }
+
+    public int getOutputBufferUsed()
+    {
+        return outputBufferUsed;
+    }
+
+    public int getInputRequired()
+    {
+        return inputRequired;
+    }
+
+    public int getRequestedOutputSize()
+    {
+        return requestedOutputSize;
+    }
+
+    public void partialDecompress(
+            final Object inputBase,
+            final long inputAddress,
+            final long inputLimit,
+            final byte[] outputArray,
+            final int outputOffset,
+            final int outputLimit)
+    {
+        if (inputRequired > inputLimit - inputAddress) {
+            throw new IllegalArgumentException(format(
+                    "Required %s input bytes, but only %s input bytes were supplied",
+                    inputRequired,
+                    inputLimit - inputAddress));
+        }
+        if (requestedOutputSize > 0 && outputOffset >= outputLimit) {
+            throw new IllegalArgumentException("Output is ready, but output buffer is full");
```

Comment:
It'd be more accurate to say "Not enough space in output buffer to output <requestedOutputSize>"